tests/integrationtests/tst_perfparser.cpp

```diff
@@ -0,0 +1,339 @@
+  Author: Nate Rogers <nate.rogers@kdab.com>
+
+  Licensees holding valid commercial KDAB Hotspot licenses may use this file in
+  accordance with Hotspot Commercial License Agreement provided with the Software.
+
+  Contact info@kdab.com if any conditions of this licensing are not clear to you.
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <QObject>
+#include <QTest>
+#include <QDebug>
+#include <QTextStream>
+#include <QProcess>
+#include <QSignalSpy>
+
+#include "perfrecord.h"
+#include "perfparser.h"
+#include "util.h"
+#include "data.h"
+#include "unistd.h"
+#include "summarydata.h"
+
+namespace {
+template<typename T>
+bool searchForChildSymbol(const T& root, const QString& searchString)
+{
+    if (root.symbol.symbol == searchString) {
+        return true;
+    } else {
+        for (auto entry : root.children) {
+            if (searchForChildSymbol(entry, searchString)) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+}
+class TestPerfParser : public QObject
+{
+    Q_OBJECT
+private slots:
+    void initTestCase()
+    {
+        qRegisterMetaType<SummaryData>();
+        qRegisterMetaType<Data::BottomUp>();
+        qRegisterMetaType<Data::TopDown>();
+        qRegisterMetaType<Data::CallerCalleeEntryMap>("Data::CallerCalleeEntryMap");
+    }
+
+    void testCppInliningNoOptions()
+    {
+        const QStringList perfOptions = {"-F", "999"};
+        QStringList exeOptions;
+
+        const QString exePath = qApp->applicationDirPath() + "/../tests/test-clients/cpp-inlining/cpp-inlining";
+        m_outputPath = qApp->applicationDirPath() + "/no-flags-cpp-in-perf.data";
+
+        perfRecord(perfOptions, exePath, exeOptions);
+
+        testPerfData("hypot", "libm", "hypot", "libm");
+    }
+
+    void testCppInliningCallGraphDwarf()
+    {
+        const QStringList perfOptions = {"--call-graph", "dwarf"};
+        QStringList exeOptions;
+
+        const QString exePath = qApp->applicationDirPath() + "/../tests/test-clients/cpp-inlining/cpp-inlining";
+        m_outputPath = qApp->applicationDirPath() + "/call-graph-cpp-in-perf.data";
+
+        perfRecord(perfOptions, exePath, exeOptions);
+
+        testPerfData("hypot", "libm", "start", "cpp-inlining");
+
+        auto bottomUpTopResult = std::max_element(m_bottomUpData.root.children.begin(), m_bottomUpData.root.children.end(),
```

Comment:
to share code, put the max_element into a free function, i.e. something like:

```
bool compareCosts(const Data::TopDown &lhs, const Data::TopDown &rhs, const TopDownResults &results)
{
    return results.inclusiveCosts.cost(0, lhs.id) < results.inclusiveCosts.cost(0, rhs.id);
}

bool compareCosts(const Data::BottomUp &lhs, const Data::BottomUp &rhs, const BottomUpResults &results)
{
    return results.costs.cost(0, lhs.id) < results.costs.cost(0, rhs.id);
}

template<typename Results>
auto maxElement(const Results &collection) -> decltype(collection.root.children.begin())
{
    using DataType = decltype(*collection.root.children.begin());
    return std::max_element(results.root.children.begin(), results.children.end(),
                                            [results](const DataType &lhs, const DataType &rhs) {
                                                return compareCosts(lhs, rhs, results);
                                            });
}
```

not tested, but I hope it's enough to get you up and running